### PR TITLE
tests: update dev environment to LLVM12

### PR DESCRIPTION
**Pull request by fbs**
As discussed in #1797 the various llvm versions required for dev is
annoying. As 12 fixes the allocation bug 12 will be our new default.

### CODE REVIEW

**Comment by fbs** [comment]
bit lost on this test failure, it's not crashing in any LLVM related code at all:

```
#0  0x00007fffeca49c5e in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#1  0x00007fffeca4ae4c in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#2  0x00007fffeca4b806 in _Unwind_Find_FDE () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#3  0x00007fffeca47848 in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#4  0x00007fffeca48a10 in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#5  0x00007fffeca48f89 in _Unwind_RaiseException () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#6  0x00007fffefe7ee4c in __cxa_throw () from /lib/x86_64-linux-gnu/libbcc.so.0
#7  0x00007fffed770941 in std::__throw_invalid_argument(char const*) ()
   from /lib/x86_64-linux-gnu/libbcc.so.0
#8  0x000055555586ddc8 in __gnu_cxx::__stoa<long long, long long, char, int> (
    __convf=0x7fffec891c20 <__strtol>, __name=0x5555559d008a "stoll", __str=0x555555d1a590 "f",
    __idx=0x7fffffffcc00) at /usr/include/c++/9/ext/string_conversions.h:83
#9  0x000055555586d344 in std::__cxx11::stoll (__str="f", __idx=0x7fffffffcc00, __base=0)
    at /usr/include/c++/9/bits/basic_string.h:6519
#10 0x00005555558ce700 in bpftrace::ast::AttachPointParser::uprobe_parser (this=0x7fffffffd110,
    allow_offset=true, allow_abs_addr=true) at /vagrant/src/ast/attachpoint_parser.cpp:404
#11 0x00005555558cd447 in bpftrace::ast::AttachPointParser::parse_attachpoint (this=0x7fffffffd110,
    ap=...) at /vagrant/src/ast/attachpoint_parser.cpp:137
#12 0x00005555558cca2b in bpftrace::ast::AttachPointParser::parse (this=0x7fffffffd110)
    at /vagrant/src/ast/attachpoint_parser.cpp:37
#13 0x000055555582103a in bpftrace::Driver::parse (this=0x7fffffffd410) at /vagrant/src/driver.cpp:56
#14 0x0000555555820e9b in bpftrace::Driver::parse_str (this=0x7fffffffd410,
    script="uprobe:/bin/sh:f { @x = func }") at /vagrant/src/driver.cpp:33
#15 0x000055555577700d in bpftrace::test::codegen::test (bpftrace=...,
    input="uprobe:/bin/sh:f { @x = func }", name="builtin_func_uprobe")
    at /vagrant/tests/codegen/common.h:45
#16 0x000055555577936e in bpftrace::test::codegen::codegen_builtin_func_uprobe_Test::TestBody (
    this=0x555555cce5e0) at /vagrant/tests/codegen/builtin_func.cpp:15
```

```
TEST(codegen, builtin_func_uprobe)
{
  auto bpftrace = get_mock_bpftrace();
  test(*bpftrace, "uprobe:/bin/sh:f { @x = func }", NAME);
}
```

```
    if (allow_abs_addr)
    {
      // First try to parse as addr. If that fails, go with symbol name.
      try
      {
        std::size_t idx;
        ap_->address = std::stoll(parts_[2], &idx, 0);

        // If there are trailing non-numeric characters, we treat the argument
        // as a string symbol. This exception will immediately be caught.
        if (idx != parts_[2].size())
        {
          ap_->address = 0;
          throw std::runtime_error("use string version");
        }
      }
      catch (const std::exception &ex)
      {
        ap_->func = parts_[2];
      }
    }
```

**Comment by fbs** [comment]
When running the uprobe test first or alone it works just fine:

```
vagrant@ubuntu-focal:~/build$ ./tests/bpftrace_test --gtest_filter=codegen.builtin_* --gtest_shuffle
Note: Google Test filter = codegen.builtin_*
Note: Randomizing tests' orders with a seed of 54710 .
[==========] Running 22 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 22 tests from codegen
[ RUN      ] codegen.builtin_func_uprobe
[       OK ] codegen.builtin_func_uprobe (13 ms)
[ RUN      ] codegen.builtin_retval
[       OK ] codegen.builtin_retval (8 ms)
[ RUN      ] codegen.builtin_ctx
```

```
Note: Google Test filter = codegen.builtin_func_uprobe
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from codegen
[ RUN      ] codegen.builtin_func_uprobe
[       OK ] codegen.builtin_func_uprobe (14 ms)
[----------] 1 test from codegen (14 ms total)
```

**Comment by fbs** [comment]
```
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 1278ee9..cb3bd16 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -16,7 +16,7 @@ add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/tests/codegen_includes.cpp
     -P ${CMAKE_SOURCE_DIR}/tests/codegen/generate_codegen_includes.cmake
   DEPENDS ${CODEGEN_SOURCES})

-if(${LLVM_VERSION} VERSION_LESS 11.0.0)
+if(${LLVM_VERSION} VERSION_LESS 13.0.0)
   set(CODEGEN_SRC ${CMAKE_BINARY_DIR}/tests/codegen_includes.cpp)
 else()
   set(CODEGEN_SRC "")
```

just this crashes it too

**Comment by fbs** [comment]
@danobi, do you have any idea? Seems like LLVM breaks the stack unwinding somehow

**Comment by fbs** [comment]
Wasn't able to find the issue. Exception handling somehow corrupts. Not sure if the eh_* sections are bad or whether it corrupts it. But it only happens in tests, haven't seen it IRL yet. 

I've added a patch to drop the exceptions and use non-throwing code instead. Its ugly but good enough IMO

**Comment by fbs** [comment]
@danobi finally ready for review

File: src/bpforc.h [hunk 1]
```diff
@@ -55,6 +55,11 @@ class MemoryManager : public SectionMemoryManager
                                StringRef SectionName,
                                bool isReadOnly) override;
 
+  ~MemoryManager()
+  {
+    deregisterEHFrames();
```
Comment (danobi) [inline src/bpforc.h:60]: I'm in awe of how you figured this out

**Review by danobi — APPROVED** [review]
Amazing work. Thanks so much for slogging through it all

File: src/bpforc.h (hunk 1 above)
Comment (fbs) [reply src/bpforc.h:60]: debuginfod on debian saved me here. Once i was certain that it was valid at some point I just set a watch point on the internal (libgcc) variable that holds them and have the backtrace tell it all. Getting the required symbols in the bionic based builds was impossible

Comment (Birch-san) [reply src/bpforc.h:60]: Amazing work! I'm gonna have to read up on the tools you used.

**Comment by Birch-san** [comment]
> todo:
>     * after build it fails with
>       ./tests/bpftrace_test --gtest_filter=codegen.*
>       : CommandLine Error: Option 'help-list' registered more than once!
>       LLVM ERROR: inconsistency in registered CommandLine options
>       Aborted (core dumped)

Hey, is this evidence that we have a way to control the Command line arguments passed to libLLVM?

I was looking for a way to do this, in order to support big strings (memset of greater than 1024 bytes):  
https://github.com/iovisor/bpftrace/discussions/1800#discussioncomment-689616

Any idea where/how we can supply command-line arguments?

**Comment by danobi** [comment]
> > todo:
> > * after build it fails with
> > ./tests/bpftrace_test --gtest_filter=codegen.*
> > : CommandLine Error: Option 'help-list' registered more than once!
> > LLVM ERROR: inconsistency in registered CommandLine options
> > Aborted (core dumped)
> 
> Hey, is this evidence that we have a way to control the Command line arguments passed to libLLVM?

I think the above error is the option framework llvm uses. I looked at the code ( https://github.com/llvm/llvm-project/blob/02286d96dbb3862a7dc5f7a2481c1f763186b259/llvm/lib/Support/CommandLine.cpp#L2397-L2401 ) and it looks like it's static registration based. So the error that @fbs was seeing was b/c we included 2 copies of the `HLOpt` in bpftrace. And then it errors out before main() is reached.

> 
> I was looking for a way to do this, in order to support big strings (memset of greater than 1024 bytes):
> [#1800 (comment)](https://github.com/iovisor/bpftrace/discussions/1800#discussioncomment-689616)
> 
> Any idea where/how we can supply command-line arguments?

I think there has to be a way. I'll try to take a quick look

**Comment by fbs** [comment]
> > > todo:
> > > 
> > > * after build it fails with
> > >   ./tests/bpftrace_test --gtest_filter=codegen.*
> > >   : CommandLine Error: Option 'help-list' registered more than once!
> > >   LLVM ERROR: inconsistency in registered CommandLine options
> > >   Aborted (core dumped)
> > 
> > 
> > Hey, is this evidence that we have a way to control the Command line arguments passed to libLLVM?
> 
> I think the above error is the option framework llvm uses. I looked at the code ( https://github.com/llvm/llvm-project/blob/02286d96dbb3862a7dc5f7a2481c1f763186b259/llvm/lib/Support/CommandLine.cpp#L2397-L2401 ) and it looks like it's static registration based. So the error that @fbs was seeing was b/c we included 2 copies of the `HLOpt` in bpftrace. And then it errors out before main() is reached.

Yes it is, see https://github.com/iovisor/bpftrace/pull/1806/commits/4c56968bb0454cd0fb3977878d505c81259427b4

**Pull request merged**